app/handlers/tests/test_handlers_services.py

```diff
@@ -0,0 +1,184 @@
+    with patch("kubernetes.client.CustomObjectsApi") as k8sclient_mock:
+        k8sclient_mock.return_value = client_mock
+        yield client_mock
+
+
+def test_service_to_twingate_resource_with_alias(example_service_body):
+    result = service_to_twingate_resource(example_service_body, "default")
+
+    assert result == {
+        "apiVersion": "twingate.com/v1beta",
+        "kind": "TwingateResource",
+        "metadata": {
+            "name": "my-service-resource",
+        },
+        "spec": {
+            "name": "my-service-resource",
+            "address": "my-service.default.svc.cluster.local",
+            "alias": "myapp.internal",
+            "protocols": {
+                "allowIcmp": False,
+                "tcp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 80, "end": 80}, {"start": 443, "end": 443}],
+                },
+                "udp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 22, "end": 22}],
+                },
+            },
+        },
+    }
+
+
+def test_service_to_twingate_resource_without_alias(example_service_body):
+    body = example_service_body
+    del body.meta["annotations"]["twingate.com/expose-alias"]
+    result = service_to_twingate_resource(body, "default")
+
+    assert result == {
+        "apiVersion": "twingate.com/v1beta",
+        "kind": "TwingateResource",
+        "metadata": {
+            "name": "my-service-resource",
+        },
+        "spec": {
+            "name": "my-service-resource",
+            "address": "my-service.default.svc.cluster.local",
+            "protocols": {
+                "allowIcmp": False,
+                "tcp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 80, "end": 80}, {"start": 443, "end": 443}],
+                },
+                "udp": {
+                    "policy": "RESTRICTED",
+                    "ports": [{"start": 22, "end": 22}],
+                },
+            },
+        },
+    }
+
+
+def test_k8s_get_twingate_resource_handles_404_returns_none(
+    k8s_customobjects_client_mock,
+):
+    k8s_customobjects_client_mock.get_namespaced_custom_object.side_effect = (
+        kubernetes.client.exceptions.ApiException(status=404)
+    )
+    assert k8s_get_twingate_resource("default", "test") is None
+
+
+def test_k8s_get_twingate_resource_reraises_non_404_exceptions(
+    k8s_customobjects_client_mock,
+):
+    k8s_customobjects_client_mock.get_namespaced_custom_object.side_effect = (
+        kubernetes.client.exceptions.ApiException(status=500)
+    )
+    with pytest.raises(kubernetes.client.exceptions.ApiException):
+        k8s_get_twingate_resource("default", "test")
+
+
+class TestTwingateServiceCreate:
+    def test_create_service(
+        self, example_service_body, kopf_handler_runner, k8s_customobjects_client_mock
+    ):
+        service_body = example_service_body
```

Comment:
nit: we probably could name the factory `service_body` to avoid this renaming here